tests/components/tasmota/test_discovery.py

```diff
@@ -0,0 +1,372 @@
+        await setup_tasmota_helper(hass)
+
+        async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/123456/configuration", "{}")
+        await hass.async_block_till_done()
+        assert not mock_tasmota_has_entities.called
+
+
+async def test_invalid_message(hass, mqtt_mock, caplog):
+    """Test receiving an invalid message."""
+    with patch(
+        "homeassistant.components.tasmota.discovery.tasmota_has_entities_with_platform"
+    ) as mock_tasmota_has_entities:
+        await setup_tasmota_helper(hass)
+
+        async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/123456/config", "asd")
+        await hass.async_block_till_done()
+        assert "Invalid discovery message" in caplog.text
+        assert not mock_tasmota_has_entities.called
+
+
+async def test_invalid_mac(hass, mqtt_mock, caplog):
+    """Test topic is not matching device MAC."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+
+    with patch(
+        "homeassistant.components.tasmota.discovery.tasmota_has_entities_with_platform"
+    ) as mock_tasmota_has_entities:
+        await setup_tasmota_helper(hass)
+
+        async_fire_mqtt_message(
+            hass, f"{DEFAULT_PREFIX}/00000049A3BA/config", json.dumps(config)
+        )
+        await hass.async_block_till_done()
+        assert "MAC mismatch" in caplog.text
+        assert not mock_tasmota_has_entities.called
+
+
+async def test_correct_config_discovery(
+    hass, mqtt_mock, caplog, device_reg, entity_reg, setup_tasmota
+):
+    """Test receiving valid discovery message."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    config["rl"][0] = 1
+    mac = config["mac"]
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+
+    # Verify device and registry entries are created
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry is not None
+    entity_entry = entity_reg.async_get("switch.beer")
+    assert entity_entry is not None
+
+    state = hass.states.get("switch.beer")
+    assert state is not None
+    assert state.name == "Beer"
+
+    assert (mac, "switch", 0) in hass.data[ALREADY_DISCOVERED]
+
+
+async def test_device_discover(
+    hass, mqtt_mock, caplog, device_reg, entity_reg, setup_tasmota
+):
+    """Test setting up a device."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    mac = config["mac"]
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+
+    # Verify device and registry entries are created
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry is not None
+    assert device_entry.manufacturer == "Tasmota"
+    assert device_entry.model == config["md"]
+    assert device_entry.name == config["dn"]
+    assert device_entry.sw_version == config["sw"]
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6285789)